File: android/app/src/main/java/com/zagir/splitty/data/AvatarStore.kt
package com.zagir.splitty.data

import android.graphics.BitmapFactory
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import com.zagir.splitty.core.network.ApiException
import com.zagir.splitty.di.ApplicationScope
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * In-memory кеш аватаров из Telegram (GET /users/{id}/avatar через бэкенд).
 * 404 («нет фото») кешируется в [missing] — без повторных походов на каждый
 * список; сетевые ошибки НЕ кешируются, попробуем при следующем показе.
 * Порт ios/Splitty/Core/AvatarStore.swift. Чистится при logout
 * (см. [OfflineDataCleaner]).
 */
@Singleton
class AvatarStore @Inject constructor(
    private val repository: SplittyRepository,
    @ApplicationScope private val scope: CoroutineScope,
) {
    private val _images = MutableStateFlow<Map<Long, ImageBitmap>>(emptyMap())

    /** Загруженные аватары по telegram user id. */
    val images: StateFlow<Map<Long, ImageBitmap>> = _images.asStateFlow()

    private val missing = mutableSetOf<Long>()
    private val inflight = mutableSetOf<Long>()

    // Поколение кеша. clear() его увеличивает, и загрузка, стартовавшая до
    // разлогина, свой результат уже не запишет: иначе ответ, пришедший через
    // секунду после logout, возвращал в кеш аватар ПРЕДЫДУЩЕГО аккаунта —
    // ровно то, что OfflineDataCleaner был должен стереть.
    private var generation = 0

    /** Запросить аватар, если его ещё нет в кеше (fire-and-forget). */
    fun request(userId: Long) {
        val started: Int
        synchronized(this) {
            if (_images.value.containsKey(userId) ||
                userId in missing || userId in inflight
            ) {
                return
            }
            inflight += userId
            started = generation
        }
        scope.launch {
            try {
                val bytes = repository.userAvatar(userId)
                val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                synchronized(this@AvatarStore) {
                    if (started != generation) return@synchronized
                    if (bitmap != null) {
                        _images.value = _images.value + (userId to bitmap.asImageBitmap())
                    } else {
                        missing += userId
                    }
                }
            } catch (e: ApiException) {
                if (e.status == 404) {
                    synchronized(this@AvatarStore) {
                        if (started == generation) missing += userId
                    }
                }
                // прочие ошибки (сеть) — не кешируем, попробуем ещё раз
            } finally {
                synchronized(this@AvatarStore) { inflight -= userId }
            }
        }
    }

    private val _fileImages = MutableStateFlow<Map<String, ImageBitmap>>(emptyMap())

    /**
     * Картинки из своего хранилища (GET /files/{id}) — фото групп. Ключ
     * строковый: id файла и id пользователя живут в разных пространствах, и
     * мешать их в одном словаре нельзя.
     */
    val fileImages: StateFlow<Map<String, ImageBitmap>> = _fileImages.asStateFlow()

    private val missingFiles = mutableSetOf<String>()
    private val inflightFiles = mutableSetOf<String>()

    /**
     * Запросить картинку по id файла. Ава неизменяема: замена даёт НОВЫЙ id,
     * поэтому кеш живёт до logout, а список групп не качает одни и те же байты
     * на каждом скролле.
     */
    fun requestFile(fileId: String) {
        val started: Int
        synchronized(this) {
            if (_fileImages.value.containsKey(fileId) ||
                fileId in missingFiles || fileId in inflightFiles
            ) {
                return
            }
            inflightFiles += fileId
            started = generation
        }
        scope.launch {
            try {
                val bytes = repository.fileData(fileId)
                val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                synchronized(this@AvatarStore) {
                    if (started != generation) return@synchronized
                    if (bitmap != null) {
                        _fileImages.value = _fileImages.value + (fileId to bitmap.asImageBitmap())
                    } else {
                        missingFiles += fileId
                    }
                }
            } catch (e: ApiException) {
                // 403/404 — файла нет или он не наш: повторять незачем.
                if (e.status == 404 || e.status == 403) {
                    synchronized(this@AvatarStore) {
                        if (started == generation) missingFiles += fileId
                    }
                }
                // прочие ошибки (сеть) — не кешируем, попробуем ещё раз
            } finally {
                synchronized(this@AvatarStore) { inflightFiles -= fileId }
            }
        }
    }

    /**
     * Забыть картинку файла: после замены или снятия авы старый кадр не должен
     * остаться на экране.
     */
    fun forgetFile(fileId: String) {
        synchronized(this) {
            _fileImages.value = _fileImages.value - fileId
            missingFiles -= fileId
        }
    }

    /** Полная очистка (logout). */
    fun clear() {
        synchronized(this) {
            generation++
            _images.value = emptyMap()
            missing.clear()
            _fileImages.value = emptyMap()
            missingFiles.clear()
            // inflight не чистим: запущенные загрузки сами снимут свои id в
            // finally, а их результат отсечёт проверка поколения.
        }
    }
}

File: ios/Splitty/Core/AvatarStore.swift
import Observation
import UIKit

/// In-memory кеш аватаров из Telegram (GET /users/{id}/avatar через бэкенд).
/// 404 («нет фото») кешируется в `missing` — без повторных походов на каждый
/// список; сетевые ошибки НЕ кешируются, попробуем при следующем показе.
/// Чистится при logout вместе с остальными пользовательскими данными.
@MainActor
@Observable
final class AvatarStore {
    private(set) var images: [Int: UIImage] = [:]
    private var missing: Set<Int> = []
    private var inflight: Set<Int> = []

    /// Поколение кеша. `removeAll()` его увеличивает, и загрузка, стартовавшая
    /// до разлогина, свой результат уже не запишет: иначе ответ, пришедший через
    /// секунду после logout, возвращал в кеш аватар ПРЕДЫДУЩЕГО аккаунта.
    private var generation = 0

    /// Загружает аватар пользователя, если он ещё не в кеше.
    func load(_ userId: Int, api: APIClient) async {
        guard images[userId] == nil, !missing.contains(userId), !inflight.contains(userId) else {
            return
        }
        let started = generation
        inflight.insert(userId)
        defer { inflight.remove(userId) }
        do {
            let data = try await api.userAvatar(id: userId)
            guard started == generation else { return }
            if let image = UIImage(data: data) {
                images[userId] = image
            } else {
                missing.insert(userId)
            }
        } catch let error as APIError {
            guard started == generation else { return }
            if case .server(let status, _, _) = error, status == 404 {
                missing.insert(userId)
            }
            // прочие ошибки (сеть) — не кешируем, попробуем ещё раз
        } catch {
            // отмена задачи и т.п. — молча
        }
    }

    /// Картинки из своего хранилища (`GET /files/{id}`) — фото групп. Ключ
    /// строковый, а не числовой: id файла и id пользователя живут в разных
    /// пространствах, и мешать их в одном словаре нельзя.
    private(set) var fileImages: [String: UIImage] = [:]
    private var missingFiles: Set<String> = []
    private var inflightFiles: Set<String> = []

    /// Загружает картинку по id файла, если её ещё нет в кеше. Ава неизменяема:
    /// замена даёт НОВЫЙ id, поэтому кеш можно держать до конца сессии, а список
    /// групп не качает одни и те же байты на каждом скролле.
    func loadFile(_ fileId: String, api: APIClient) async {
        guard fileImages[fileId] == nil,
              !missingFiles.contains(fileId),
              !inflightFiles.contains(fileId) else {
            return
        }
        let started = generation
        inflightFiles.insert(fileId)
        defer { inflightFiles.remove(fileId) }
        do {
            let data = try await api.fileData(id: fileId)
            guard started == generation else { return }
            if let image = UIImage(data: data) {
                fileImages[fileId] = image
            } else {
                missingFiles.insert(fileId)
            }
        } catch let error as APIError {
            guard started == generation else { return }
            // 403/404 — файла нет или он не наш: повторять незачем. Сетевые
            // ошибки не кешируем, попробуем ещё раз.
            if case .server(let status, _, _) = error, status == 404 || status == 403 {
                missingFiles.insert(fileId)
            }
        } catch {
            // отмена задачи и т.п. — молча
        }
    }

    /// Забыть картинку файла: после замены или снятия авы старый кадр не должен
    /// остаться на экране.
    func forgetFile(_ fileId: String) {
        fileImages[fileId] = nil
        missingFiles.remove(fileId)
    }

    /// Полная очистка (logout).
    func removeAll() {
        generation += 1
        images = [:]
        missing = []
        fileImages = [:]
        missingFiles = []
    }
}
